src/Temporalio/Worker/WorkflowReplayer.cs

```diff
@@ -0,0 +1,270 @@
+        /// <returns>Result of the replay run.</returns>
+        public async Task<WorkflowReplayResult> ReplayWorkflowAsync(
+            WorkflowHistory history, bool throwOnReplayFailure = true) =>
+            (await ReplayWorkflowsAsync(new[] { history }, throwOnReplayFailure).ConfigureAwait(false)).Single();
+
+        /// <summary>
+        /// Replay multiple workflows from the given histories.
+        /// </summary>
+        /// <param name="histories">Histories to replay.</param>
+        /// <param name="throwOnReplayFailure">If true, which is not the default, this will throw
+        /// <see cref="InvalidWorkflowOperationException" /> on a workflow task failure
+        /// (e.g. non-determinism) as soon as it's encountered. This has no effect on workflow
+        /// failures which are not reported as part of replay.</param>
+        /// <returns>Results of the replay runs.</returns>
+        public async Task<IEnumerable<WorkflowReplayResult>> ReplayWorkflowsAsync(
+            IEnumerable<WorkflowHistory> histories, bool throwOnReplayFailure = false)
+        {
+            // We could stream the results, but since the method wants them all anyways, a list is
+            // ok
+            using (var runner = new WorkflowHistoryRunner(Options, throwOnReplayFailure))
+            using (var cts = new CancellationTokenSource())
+            {
+                var workerTask = Task.Run(() => runner.RunWorkerAsync(cts.Token));
+                try
+                {
+                    var results = new List<WorkflowReplayResult>();
+                    foreach (var history in histories)
+                    {
+                        results.Add(await runner.RunWorkflowAsync(history).ConfigureAwait(false));
+                    }
+                    return results;
+                }
+                finally
+                {
+                    // Cancel and wait on complete
+                    cts.Cancel();
+                    await workerTask.ConfigureAwait(false);
+                }
+            }
+        }
+
+#if NETCOREAPP3_0_OR_GREATER
+        /// <summary>
+        /// Replay multiple workflows from the given histories.
+        /// </summary>
+        /// <param name="histories">Histories to replay.</param>
+        /// <param name="throwOnReplayFailure">If true, which is not the default, this will throw
+        /// <see cref="InvalidWorkflowOperationException" /> on a workflow task failure
+        /// (e.g. non-determinism) as soon as it's encountered. This has no effect on workflow
+        /// failures which are not reported as part of replay.</param>
+        /// <param name="cancellationToken">Cancellation token to stop replaying.</param>
+        /// <returns>Results of the replay runs.</returns>
+        public async IAsyncEnumerable<WorkflowReplayResult> ReplayWorkflowsAsync(
+            IAsyncEnumerable<WorkflowHistory> histories,
+            bool throwOnReplayFailure = false,
```

Comment:
Why do you need this flag?
I removed it in TS since the user can decide to throw while iterating throw the results.

Reply:
Suggesting that single replay throws and multiple (async iterator) doesn't.
We can give the option for the multi replay method that replays synchronously.

Reply:
> Suggesting that single replay throws and multiple (async iterator) doesn't.

That's already the case by default. Are you saying both need to disallow changing the default? Or just one have the option of changing the default.

> We can give the option for the multi replay method that replays synchronously.

They all replay synchronously in this sense. There is no concurrent replay.

Reply:
I don't think we need the option in the async iterator method.

> They all replay synchronously in this sense. There is no concurrent replay.

I was talking about the multi replay method that returns `task<iterator>`.
I'm of course familiar with how the replayer works.

Reply:
> I don't think we need the option in the async iterator method.

This argues for inconsistency. Why is a `foreach` ok to throw but not a `await foreach`?

Reply:
Because in the `task<iterator>` method you can't get the results one by one, and it's nice to be able to fail fast.
In the async iterator method, the user has full control and can abort whenever they like.

Reply:
> Because in the `task<iterator>` method you can't get the results one by one

You can. The reason for the async iterator in this case is because I allow async iteration input. You can have either:

```csharp
foreach (var result in await ReplayWorkflowsAsync(syncHistIter))
{
    // Process 1 by t1
}
```
or
```csharp
await foreach (var result in ReplayWorkflowsAsync(asyncHyncHistIter))
{
    // Process 1 by 1
}
```